docs/docs/concepts/how-it-works/spfx-project-upgrade.mdx

```diff
@@ -0,0 +1,227 @@
+
+A finding contains the information about the rule that triggered it, the recommended resolution (ie. what users need to do to apply the missing change), how the resolution should be executed (eg. a shell command, JSON modification, etc.), and whether the change is recommended or required.
+
+Additionally, each finding has one or more **finding instances**.
+
+### Finding instance
+
+A finding instance describes the name of the file, and the location in the file where the finding's resolution should be applied.
+
+Typically, each finding has one finding instance, but there are some rules that are related to code changes that need to be applied multiple times, which can lead to a finding having multiple instances.
+
+### Report
+
+When the `spfx project upgrade` command ends its execution, it presents the user with a report. The report contains the list of changes that need to be applied to the project to upgrade it to the desired SPFx version (findings).
+
+The `spfx project upgrade` command offers several types of reports, each optimized for a different scenario:
+
+- **JSON** with structured output, recommended for automated processing
+- **plain-text** with basic human-readable output
+- **markdown** with rich human-readable output, recommended for most cases
+- **CodeTour** for interactive upgrade guidance using the VSCode CodeTour plugin, recommended if you want to learn more about the changes and easily navigate through your project
+
+Report types are supported through command output in the `spfx project upgrade` command.
+
+### Project model
+
+To speed up the upgrade process, the `spfx project upgrade` command starts the upgrade by collecting the relevant project files and building an object model out of them. That way, files need to be parsed only once and rules can be applied on in-memory objects that represent the project. This makes the upgrade process fast, even when it spans multiple versions with many different changes.
+
+Project model is stored in `src/m365/spfx/commands/project/project-model`.
+
+### Exit codes
+
+Shell executables (CLIs, scripts, etc.) communicate to the parent process whether their execution succeeded using exit codes. An exit code of 0 means success and a non-0 exit code means failure. The `spfx project upgrade` command uses exit codes to communicate different error states to the parent process using it. These error codes have been originally designed for use with the PnP Yeoman Generator.
+
+The exit codes of the `spfx project upgrade` command are defined in the `src/m365/spfx/commands/project/project-upgrade.ts` file.
+
+## How it works
+
+This section explains the process that the `spfx project upgrade` command follows to produce steps to upgrade the SPFx project to the specified SPFx version.
+
+![Diagram illustrating the process of the spfx project upgrade command](../../images/how-it-works/spfx-project-upgrade-process.png)
```

Comment:
Would it be possible to have a different image here in dark mode? Currently, this hurts my eyes a bit 😄 

Another suggestion would be to scale this image down a bit as it's quite a big image to view on the page. Could we scale it down so it can be fully visible on the screen without the need or scrolling?

Reply:
Is there a way for us to switch images in the content between the two modes?

Reply:
Just had a quick look and there appears to be a component named `ThemedImage` we can use. Ref: https://docusaurus.io/docs/markdown-features/assets#themed-images

Reply:
Do you know by any chance if Docusaurus supports mermaid diagrams? It would make it easier for us to maintain it long-term as it's just code and perhaps would give us the ability to render light- and dark-mode without keeping two different images.

Here's the mermaid flowchart that we could use:

```mermaid
flowchart TD
    start(start) --> isSpfx{Executed on an SPFx project?}
    isSpfx -- no, exit: 1 --> stop(end)
    isSpfx -- yes --> isToVersionSupported{Is the version to upgrade to supported?}
    isToVersionSupported -- no, exit: 2 --> stop
    isToVersionSupported -- yes --> detectVersion(Detect SPFx version)
    detectVersion --> isVersionDetected{Version detected?}
    isVersionDetected -- no, exit: 3 --> stop
    isVersionDetected -- yes --> isFromVersionSupported{Is the from version supported?}
    isFromVersionSupported -- no, exit: 4 --> stop
    isFromVersionSupported -- yes --> isToVersionLarger{Is the to version larger than from version?}
    isToVersionLarger -- no, exit: 5 --> stop
    isToVersionLarger -- yes --> areToAndFromTheSame{Are the from and to versions the same?}
    areToAndFromTheSame -- yes --> stop
    areToAndFromTheSame -- no --> collect(Collect project)
    collect --> deltas(Calculate deltas)
    deltas --> applyRules(Apply upgrade rules)
    applyRules --> dedupe(Deduplicate findings)
    dedupe --> removeSuperseded(Remove superseded findings)
    removeSuperseded --> flatten(Flatten findings)
    flatten --> replaceTokens(Replace tokens in resolutions)
    replaceTokens --> buildReport(Build the report)
    buildReport --> stop
```

Reply:
That's a great idea. Just had a quick glance and it appears that we can enable it https://docusaurus.io/docs/markdown-features/diagrams

Reply:
Should we add it to this PR or would you rather like to keep it separate? If we add this to this PR, you could submit a commit with the necessary changes to this PR

Reply:
I've updated the PR with support for mermaid